lib/anno_utils.pm

```diff
@@ -0,0 +1,2917 @@
+        });
+    my $req = HTTP::Request->new(POST => $url);
+    $req->header('content-type', 'application/json');
+    $req->content($content);
+    my $ua = LWP::UserAgent->new();
+    
+    my $ret = $ua->request($req);
+    if (!$ret->is_success()) {
+        print("Error body from Relation Engine on NCBI taxa query:\n" .
+            $ret->decoded_content({'raise_error' => 1}));
+        # might want to try to parse content to json and extract error, try this for now
+        die "Error contacting Relation Engine: " . $ret->status_line();
+    }
+    my $retjsonref = $ret->decoded_content({'raise_error' => 1, 'ref' => 1});
+    my $retjson = decode_json($retjsonref);
+
+    if (!$retjson->{count}) {
+        die "No result from Relation Engine for NCBI taxonomy ID " . $tax_id;
+    }
+    return $retjson->{'results'}[0]{'scientific_name'};
+}
+
+#
+## get the genome data via GenomeAnnotationAPIClient
+#
+sub _get_genome {
+    my ($self, $gn_ref) = @_;
+    my $ga_client = new installed_clients::GenomeAnnotationAPIClient($self->{call_back_url});
+    my $output = $ga_client->get_genome_v1({
+        genomes => [{
+            "ref" => $gn_ref
+        }],
+        ignore_errors => 1,
+        no_data => 0,
+        no_metadata => 0,
+        downgrade => 0
+    });
+    my $genome = $output->{genomes}->[0];
+    $genome->{data}->{'_reference'} = $genome->{info}->[6]."/".$genome->{info}->[0]."/".$genome->{info}->[4];
+    my $ftr_cnt = @{$genome->{data}->{features}};
+    print("Genome $genome->{data}->{'_reference'} downloaded with $ftr_cnt features.\n");
+    return $genome->{data};
+}
+
+#
+## get the genome gff_contents
+#
+sub _get_genome_gff_contents {
+    my ($self, $obj_ref) = @_;
+
+    return [] unless ($self->_validate_KB_objref($obj_ref));
+
+    my $gff_filename = catfile($self->{genome_dir}, 'genome.gff');
+    my ($gff_contents, $attr_delimiter) = ([], "=");
+
+    # getting gff_contents from $obj_ref if it is of genome type
+    my $obj_info = $self->_fetch_object_info($obj_ref);
+
+    my $in_type = $obj_info->[2];
+    my $is_genome = ($in_type =~ /KBaseGenomes.Genome/ ||
+                     $in_type =~ /KBaseGenomeAnnotatioVns.GenomeAnnotation/);
+
+    if ($is_genome) {
+        # generating the gff file directly from genome
+        $gff_filename = $self->_write_gff_from_genome($obj_ref);
+        unless (-s $gff_filename) {
+            croak "**rast_genome ERROR: GFF file is empty!\n";
+        }
+        ($gff_contents, $attr_delimiter) = $self->_parse_gff(
+											$gff_filename, $attr_delimiter);
+    }
+	return $gff_contents;
+}
+
+
+sub _get_contigs_from_fastafile {
+    my ($self, $fpath) = @_;
+
+    my @contigs;
+    my $contigID_hash = {};
+    my $fasta = "";
+    my $fh = $self->_openRead($fpath);
+    while (my $line = <$fh>) {
+        $fasta .= $line;
+    }
+    close($fh);
+    $fasta =~ s/\>([^\n]+)\n/>$1\|\|\|/g;
+    $fasta =~ s/\n//g;
+    my $array = [split(/\>/,$fasta)];
```

Comment:
unless there is a compelling reason to use an arrayref, I would use an array here -- it is faster than using an arrayref and you don't have to worry about the reference notation.